onmt/dynamic/corpus.py

```diff
@@ -0,0 +1,228 @@
+                self.cid, report_msg))
+
+    def _add_index(self, stream):
+        for i, item in enumerate(stream):
+            example = item[0]
+            line_number = i * self.stride + self.offset
+            example['indices'] = line_number
+            if (len(example['src']) == 0 or len(example['tgt']) == 0 or
+                    ('align' in example and example['align'] == 0)):
+                # empty example: skip
+                empty_msg = f"Empty line exists in {self.cid}#{line_number}."
+                if self.skip_empty_level == 'error':
+                    raise IOError(empty_msg)
+                elif self.skip_empty_level == 'warning':
+                    logger.warning(empty_msg)
+                continue
+            yield item
+
+    def _iter_corpus(self):
+        corpus_stream = self.corpus.load(
+            stride=self.stride, offset=self.offset)
+        tokenized_corpus = self._tokenize(corpus_stream)
+        transformed_corpus = self._transform(tokenized_corpus)
+        indexed_corpus = self._add_index(transformed_corpus)
+        yield from indexed_corpus
+
+    def __iter__(self):
+        if self.infinitely:
+            while True:
+                _iter = self._iter_corpus()
+                yield from _iter
+        else:
+            yield from self._iter_corpus()
+
+
+def build_corpora_iters(corpora, transforms, corpora_info, is_train=False,
+                        skip_empty_level='warning', stride=1, offset=0):
+    """Return `ParallelCorpusIterator` for all corpora defined in opts."""
+    corpora_iters = dict()
+    for c_id, corpus in corpora.items():
+        c_transform_names = corpora_info[c_id].get('transforms', [])
+        corpus_transform = [transforms[name] for name in c_transform_names]
+        transform_pipe = TransformPipe.build_from(corpus_transform)
+        logger.info(f"{c_id}'s transforms: {str(transform_pipe)}")
+        corpus_iter = ParallelCorpusIterator(
+            corpus, transform_pipe, infinitely=is_train,
+            skip_empty_level=skip_empty_level, stride=stride, offset=offset)
+        corpora_iters[c_id] = corpus_iter
+    return corpora_iters
+
+
+def save_transformed_sample(opts, transforms, n_sample=3, build_vocab=False):
+    """Save transformed data sample as specified in opts."""
+
+    if n_sample == -1:
+        logger.info(f"n_sample={n_sample}: Save full transformed corpus.")
+    elif n_sample == 0:
+        logger.info(f"n_sample={n_sample}: no sample will be saved.")
+        return
+    elif n_sample > 0:
+        logger.info(f"Save {n_sample} transformed example/corpus.")
+    else:
+        raise ValueError(f"n_sample should >= -1, get {n_sample}.")
+
+    from onmt.dynamic.iterator import DatasetAdapter
+    corpora = get_corpora(opts, is_train=True)
+    if build_vocab:
+        counter_src = Counter()
+        counter_tgt = Counter()
+    datasets_iterables = build_corpora_iters(
+        corpora, transforms, opts.data, is_train=True,
```

Comment:
is_train=True will mean that self.infinitely will be set to True as well, right?
Wouldn't that cause an infinite loop in the vocab building with n_sample = -1 ?

Reply:
Fixed in #24.